### PR TITLE
Rating-based rank tiers: named, colored usernames (#73)

### LINKED ISSUE #73: Rating-based rank tiers: named, colored handles (Codeforces/CodeChef-style)
## Summary

Currently every username renders in the same flat link color across the platform. Codeforces and CodeChef both give competitors a **named rank** and a **color** derived from their rating — an at-a-glance signal of skill. serious-oj already runs a Codeforces-style Elo rating system (stored per-domain on each user), so the data exists; it's just not surfaced visually.

This issue tracks introducing a **rating-based rank tier system**: each user's absolute rating maps to a named tier with its own color, and the username text is colored accordingly everywhere it appears.

## Proposed tiers (platform-unique "Achievement title" theme)

| Rank | Rating range | Color |
|------|--------------|-------|
| Novice | 0–199 | gray `#7d8590` |
| Apprentice | 200–399 | green `#2f9e44` |
| Specialist | 400–599 | teal `#0c8599` |
| Expert | 600–799 | blue `#1c6fd6` |
| Master | 800–999 | violet `#7048e8` |
| Elite | 1000–1199 | orange `#e8820c` |
| Legend | 1200+ | red `#e03131` |
| Unrated | no rated contest | default link color |

> **Note on thresholds (updated 2026-05-25).** The original Codeforces-shape thresholds (Legend at 2600) assumed a rating range up to ~3800. Live data on this instance has top users sitting under 900, which would render every existing user as Novice and make the tier system meaningless. Thresholds were rescaled to 200-point bands with Legend at 1200 so today's top user lands in Master with room to grow. Implemented in #74, commit `7bac11c`.

## Scope

- Single source of truth for tier definitions + a `get_rating_rank()` helper in the user model layer.
- Colored handle text **everywhere** a username renders (ranking, scoreboards, discussions, comments, record detail, sidebars, nav bar, profile header).
- Tier **name** displayed on the user profile page.
- **Replace** the existing `LV 0`–`LV 10` percentile badge — the colored handle becomes the single rank indicator (the percentile badge measured the same rating, just relatively, and two color-coded markers on one name is redundant/noisy).
- No change to the Elo algorithm or stored rating values. No database migration.

## Out of scope

- No tier badge/star icon, no hover tooltips (colored text only).
- The backend `level` field computation in the ranking job is left intact (unused) — full removal is a separate follow-up.

## References

- Codeforces rank colors: https://codeforces.com/blog/entry/20638
- CodeChef star rating: https://www.codechef.com/ratings

### PR DESCRIPTION
**Pull request by cloud-007**
## Summary

Replaces the flat, uniform username color with a **rating-based rank tier** system in the spirit of Codeforces and CodeChef. Each user's Elo rating maps to a named tier with its own color, and the username text is colored accordingly across the whole platform.

Closes #73.

### Tiers

| Rank | Rating | Color |
|------|--------|-------|
| Novice | 0–199 | gray `#7d8590` |
| Apprentice | 200–399 | green `#2f9e44` |
| Specialist | 400–599 | teal `#0c8599` |
| Expert | 600–799 | blue `#1c6fd6` |
| Master | 800–999 | violet `#7048e8` |
| Elite | 1000–1199 | orange `#e8820c` |
| Legend | ≥ 1200 | red `#e03131` |
| Unrated | no rated contest | default link color |

Thresholds use 200-point bands with Legend at 1200. The original Codeforces-shape thresholds (Legend at 2600) assumed a rating range up to ~3800; with actual top users on this instance sitting under 900, every existing user rendered as Novice, making the tier system meaningless for our data. The rescaled bands put today's top user in Master with room to grow.

### What changed

- **Single source of truth** — `RATING_RANKS` + `get_rating_rank()` added to `vj4/model/builtin.py`. Pure function, fully unit-tested.
- **Colored handles everywhere** — the shared `render_inline` user macro applies a `user-rated--<slug>` class, so ranking, scoreboards, discussions, comments, record list & detail, sidebars all get colored handles. Plus the nav bar and the profile page heading.
- **Tier name on the profile page** — the profile stats line now reads e.g. `Expert · Solved 12 problems, (No. 3), Rating: 1820`.
- **`LV 0`–`LV 10` percentile badge removed** — it was derived from the same rating, so a second color-coded marker on the handle was redundant. The colored handle is now the single rank indicator. The `MOD` badge is untouched. The backend `level`/`LEVELS` computation in `rank.py` is intentionally left intact (unused) — full removal is a separate follow-up.
- No database migration, no change to the Elo rating algorithm.

### Regression caught and fixed during verification

The rating lookup initially used `dudoc.get('rating')` directly. While verifying, a 500 was reproduced on `record_detail.html` and `user_detail.html` for any user with no domain-user document — `domain.get_user()` returns `None`, and `None.get(...)` raises `UndefinedError` in Jinja. Fixed by guarding the call: `dudoc.get('rating') if dudoc else none` (commit `05f7434`). The nav bar uses `handler.domain_user`, which `vj4/handler/base.py` guarantees is always a dict, so it needs no guard.

## Test Plan

- [x] `python -m pytest vj4/test/test_builtin.py` — 5/5 pass against the new thresholds (boundary values, `None`, negative, very-high ratings)
- [x] `node scripts/build` — exits 0, all 7 `user-rated--*` classes present in the output CSS
- [x] All three changed templates compile against serious-oj's real Jinja environment
- [x] `render_inline` macro rendered with rated / unrated / `None` / `{}` `dudoc` — correct class emitted, no crash
- [ ] **Manual:** open the site and eyeball nav-bar contrast — confirm the seven tier colors read legibly against the nav background (if any tier reads poorly, it is a one-line tweak in `\$rating-color`)

### CODE REVIEW

**Comment by cloud-007** [comment]
Converting this to a **draft — please do not merge yet.**

The implementation and automated checks are complete (unit tests, build, template render checks), but we want to finish a round of manual browser verification — handle colors across all tiers, nav-bar contrast, the ranking page — and apply any fixes that surface before this is merged.

Will move it back to *ready for review* once verified.

**Comment by cloud-007** [comment]
Marking this **ready for review**.

Since it was drafted, the Codeforces-style profile header refinement was added (6 commits, each spec- and code-reviewed):
- rank-title line above the handle, and a rating line directly under it showing current rating + peak/max rating
- dark scrim panel behind the header text, fixing colored-handle contrast against the photo background
- new `get_user_max_rating` model function (unit-tested) feeding the max rating to the profile

All model tests pass, the frontend build is clean, and rated/unrated profiles render correctly.

Note: a 500 seen on `/ranking` during local testing was a **pre-existing data issue** — an orphan `domain.user` row for a sentinel uid (-1) that has no `user` document. `ranking_main.html` is not touched by this PR. Hardening the ranking page against such rows is worth a small separate PR.

File: vj4/ui/templates/user_detail.html [hunk 1]
```diff
@@ -11,15 +11,30 @@
...
                 {% else %}
                   {{ udoc['uname'] }}
                 {% endif %}
+                </span>
                 <small>({{ _('UID') }}: {{ udoc['_id'] }})</small>
               </h1>
+              {% if _rank %}
+                <p class="profile-header__rating">
+                  {{ _('Rating') }}: {{ dudoc['rating'] }}
+                  {% if max_rating %}
```
Comment (ajami1331) [inline vj4/ui/templates/user_detail.html]:
```suggestion
                  {% if max_rating is not none %}
```

**Review by ajami1331 — CHANGES_REQUESTED** [review]

File: vj4/ui/templates/user_detail.html (hunk 1 above)
Comment (cloud-007) [reply vj4/ui/templates/user_detail.html]: Applied in 09f3632.

Worth noting: in practice this line is unreachable today — `max_rating` is sourced from `rating_changes.new_rating`, and the Elo job's binary search range is 1–8000, so the field is either `None` or a positive int (never 0). Both forms behave identically against today's data. The explicit `is not none` is still the right call: it expresses the intent ("we have a value to show") and removes a latent footgun if the rating floor ever changes.

**Comment by cloud-007** [comment]
@ajami1331 re-review please — your suggestion is applied in 09f3632, master merged in 561d618, and the review thread is resolved. Branch is no longer behind.

**Review by ajami1331 — APPROVED** [review]
LGTM
